azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +62,166 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
+            GoalState._Certs = Certificates(xml_text)
 
-            uri = findtext(xml_doc, "SharedConfig")
-            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-            self.shared_conf = SharedConfig(xml_text)
+        self.certs = GoalState._Certs
 
-            uri = findtext(xml_doc, "Certificates")
-            if uri is None:
-                self.certs = None
-            else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
-                self.certs = Certificates(xml_text)
-
-            uri = findtext(xml_doc, "ExtensionsConfig")
-            if uri is None:
-                self.ext_conf = ExtensionsConfig(None)
-            else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-                self.ext_conf = ExtensionsConfig(xml_text)
+        GoalState.ContainerID = self.container_id
 
-            uri = findtext(container, "RemoteAccessInfo")
-            if uri is None:
-                self.remote_access = None
-            else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
-                self.remote_access = RemoteAccess(xml_text)
-        except Exception as e:
-            logger.warn("Fetching the goal state failed: {0}", ustr(e))
-            raise
-        finally:
-            logger.info('Fetch goal state completed')
+        # The changed property on GoalState indicates only if the incarnation changed
+        # There's also a changed property on ext_config that indicates if they changed
+        self.changed = False
+        if current_goal_state is None:
+            self.changed = True
+        elif current_goal_state is not None and self.incarnation != current_goal_state.incarnation:
+            self.changed = True
 
     @staticmethod
-    def fetch_goal_state(wire_client):
+    def fetch_goal_state(wire_client, ext_config_retriever, current_goal_state=None):
         """
         Fetches the goal state, not including any nested properties (such as extension config).
         """
-        return GoalState(wire_client)
-
-    @staticmethod
-    def fetch_full_goal_state(wire_client):
-        """
-        Fetches the full goal state, including nested properties (such as extension config).
-        """
-        return GoalState(wire_client, full_goal_state=True)
+        return GoalState(wire_client, ext_config_retriever, current_goal_state)
 
-    @staticmethod
-    def fetch_full_goal_state_if_incarnation_different_than(wire_client, incarnation):
-        """
-        Fetches the full goal state if the new incarnation is different than 'incarnation', otherwise returns None.
-        """
-        goal_state = GoalState(wire_client, base_incarnation=incarnation)
-        return goal_state if goal_state.incarnation != incarnation else None
+    @property
+    def hosting_env(self):
+        if not self._hosting_env_retrieved:
```

Comment:
I'm sorry I don't understand the need for an extra variable (`_hosting_env` vs `_hosting_env_retrieved`). Isin't a None check enough? Why do we need an extra variable?

Reply:
If hosting environment isn't present in the goal state, then _hosting_env will be None and _hosting_env_retrieved will be True. If we just using _hosting_env, then we'll keep trying to retrieve it if it's not in the goal state.